src/utils/logger-custom-format.js

```diff
@@ -0,0 +1,23 @@
+const { format } = require('winston');
+const stack = require('callsite');
+const _ = require('lodash');
+
+const customFormat = () => {
+  const { printf } = format;
+
+  return printf((info) => {
+    const { level, message } = info;
+
+    if (info instanceof Error) {
+      const site = _.get(stack(), [9]);
+      return `${level}: ${site.getFunctionName() || 'anonymous'} ${site.getFileName()} ${site.getLineNumber()}`;
+    }
+
+    const { req } = message;
+    const { url, method } = req;
+    const currentTimestamp = (new Date()).toISOString();
+    return `${currentTimestamp} ${level}: ${method} ${url}`;
+  });
+};
+
```

Comment:
This is a good start. However, the implementation is tightly coupled with the `express` request object. What happens when you want to log an error function in your application that is not a route controller and does not have a request object?

Reply:
The error logger is not tightly coupled with the express request object. In fact, it does not rely on the request object. 

I think you are getting confused by these lines line 16-19 in `logger-custom-format.js`:
```
    const { req } = message;
    const { url, method } = req;
    const currentTimestamp = (new Date()).toISOString();
    return `${currentTimestamp} ${level}: ${method} ${url}`;
```

They are being used by the the middleware logRequest in `log-request.js` which I used to print info about the incoming request.

Reply:
Try using the error-logger outside of a route controller and you'll see the tight coupling. For example:
1. go to _src/index.js_
2. import the logger (`const errorLogger = request('./utils/error-logger');`)
3. log an error anywhere in that file
```
try {
  invalid-javascript
} catch(e) {
  errorLogger.error(e);
}
```

Reply:
I tried using:

```
try {
  throw new Error('Error thrown from index.js');
} catch (e) {
  errorLogger.error(e);
}
```

and when I start the app, I get this:
```
> shopping-site-reactjs-backend@1.0.0 start C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend
> node src

error: anonymous C:\Users\Antonio Lok\Documents\workspace\online-shopping-reactJS-Backend\src\index.js 43
Running on localhost:8000
We are connected
(node:22936) DeprecationWarning: collection.ensureIndex is deprecated. Use createIndexes instead.
```

Reply:
I'm sorry, my feedback was poor. The example that I used to test the code wasn't the example that I shared in my comment. What I tested the code with was:

```
errLogger.error('Something when wrong');
```
which threw an error:
```
~/online-shopping-reactJS-Backend/node_modules/winston/lib/winston/logger.js:307
      throw ex;
      ^

TypeError: Cannot destructure property `url` of 'undefined' or 'null'.
```

When using the error logger, I assumed that as an engineer I can use it to replace `console.log`. But it seems that your code only expects an error object or an express `request` object, which is why I mentioned that your code is tightly coupled to with express.

Given this finding, what are your thoughts on how this should be?

Reply:
Ah okay, that makes more sense now. Thanks for clarifying, now I get why you said that my error logger utility is somewhat dependent on the request object. You have probably noticed this already but this happens because I "expected" only error objects to be passed to the error logger function. Your example passes a non-error object and so my if statement `    if (info instanceof Error) {` is not triggered and it executes the following lines which relies on the req object. "One" solution I could think was to separate the logic into 2 different format functions. 1 used by the error logger function (which now does not depend on an error object passed) and the other one is used by the middleware which is tightly coupled to the req object.

Reply:
Right. But you only looking at use cases that present themselves to you. You've created a logger that either takes an error object or an express request object.

Without over-engineering a solution, I would encourage you to step back and think about all the **reasonable** use cases that you may need in the _immediate_ or _near future_. Otherwise, you will end up refactoring your work too often, which is not so efficient. I'm finding that you're trying to solve _only_ the problem that is right in front of you. It is sometimes better to start slow in the beginning so you can move fast in the end.

Reply:
"You've created a logger that either takes an error object or an express request object."

Actually.... I have created a logger which takes anything: error objects, arrays, strings, etc. I am not sure why you would say "either" an error object or request object. I remember testing my code with your example which is a string being passed to the logger:
`
logger.error('Something when wrong'); // A string
`
and that worked fine. I have also tested it with arrays.

_________________________________________

"and think about all the reasonable use cases that you may need in the immediate or near future." 

I think I know which "use cases" you mean. I decided before to not include those use cases since they are not being used anywhere right now. I am not a huge fan of adding functionality or code which is not being used at the current moment. But you are right about "you will end up refactoring your work too often". I have refactored a lot of code and implementation due to this. I guess this is something that I need to work on more...
Nevertheless, I have modified the log message to include the message being passed as well as making it look a bit nicer. I have also added support to log messages of level info, warn and error. I have updated the PR to provide some examples on how to test this and sample outputs.

Reply:
I agree you don't want to over-engineer your solution. So, it's definitely a balance to decide "will I need this functionality in the near future". I think it is indeed reasonable to assume that you will need that functionality in the near future 👍

Reply:
Haha yup. Thanks a lot for the review, I have said this and I will say it again, I really appreciate your time and help ^-^